README.rst
Redkyn-common |pypi| |ci|
=========================

.. |pypi| image:: https://badge.fury.io/py/redkyn-common.svg
    :target: https://badge.fury.io/py/redkyn-common

.. |ci| image:: https://github.com/redkyn/redkyn-common/workflows/Build/badge.svg?branch=master&event=push
    :target: https://github.com/redkyn/redkyn-common/actions?query=workflow%3A%22Build%22
    :alt: Build Status

Common libraries for `Assigner <https://github.com/redkyn/assigner>`_ and `Grader <https://github.com/redkyn/grader>`_.

Libraries
---------
- ``canvas``: An API wrapper for getting course information from Canvas


CHANGELOG.md
## Devel

- Official support for Python 3.9

## 1.1.0

- Add support for fetching course assignments from Canvas
- Add support for pushing assignment scores to Canvas

## 1.0.1

- Add dependency on `typing` for python 3.4 support

## 1.0.0

- Add Canvas API


redkyn/tests/utils.py
import unittest
from unittest.mock import patch


class TestCase(unittest.TestCase):
    @classmethod
    def tearDownClass(cls):
        # pylint: disable=C0321
        if hasattr(cls, "_CLASS_CLEANUP"):
            for x in cls._CLASS_CLEANUP:
                x()

    @classmethod
    def _create_class_patch(cls, target, **kwargs):
        """
        Shortcut for creating a class-level patch with proper
        cleanup handled.
        """
        target_patch = patch(target, **kwargs)
        target_mock = target_patch.start()

        if not hasattr(cls, "_CLASS_CLEANUP"):
            cls._CLASS_CLEANUP = []

        cls._CLASS_CLEANUP.append(target_patch.stop)
        return target_mock

    def _create_patch(self, target, **kwargs):
        """
        Shortcut for creating a class and having it cleaned up
        properly.
        """
        target_patch = patch(target, **kwargs)
        target_mock = target_patch.start()
        self.addCleanup(target_patch.stop)

        return target_mock

class IntegrationTestCase(TestCase):
    def setUp(self):
        self.mock_logging = self._create_patch(
            "assigner.logging", autospec=True
        )


redkyn/tests/canvas_tests.py
from redkyn.tests.utils import TestCase

from redkyn.canvas import CanvasAPI
from redkyn.canvas.exceptions import AuthenticationFailed, CourseNotFound, NameResolutionFailed

from requests.exceptions import HTTPError

class FakeResponse:
    def __init__(self, status_code):
        self.status_code = status_code


class CanvasTestCase(TestCase):
    def setUp(self):
        self.mock_requests = self._create_patch("redkyn.canvas.requests", autospec=True)
        self.mock_get = self._create_patch("redkyn.canvas.requests.get")

    def test_urls_are_https(self):
        api = CanvasAPI("test token", "mst.instructure.com")
        self.assertEqual("https://mst.instructure.com", api.website_root)

    def test_http_urls_are_httpsified(self):
        api = CanvasAPI("test token", "http://mst.instructure.com")
        self.assertEqual("https://mst.instructure.com", api.website_root)

    def test_courses_auth_failure(self):
        self.mock_get.side_effect = HTTPError(response=FakeResponse(401))

        api = CanvasAPI("test token", "mst.instructure.com")
        with self.assertRaises(AuthenticationFailed):
            api.get_instructor_courses()

    def test_students_auth_failure(self):
        self.mock_get.side_effect = HTTPError(response=FakeResponse(401))

        api = CanvasAPI("test token", "mst.instructure.com")
        with self.assertRaises(AuthenticationFailed):
            api.get_course_students(420)

    def test_course_not_found(self):
        self.mock_get.side_effect = HTTPError(response=FakeResponse(404))

        api = CanvasAPI("test token", "mst.instructure.com")
        with self.assertRaises(CourseNotFound):
            api.get_course_students(420)

    def test_failed_dns(self):
        self.mock_get.side_effect = HTTPError("Name or service not known")

        api = CanvasAPI("test token", "mst.instructure.com")

        with self.assertRaises(NameResolutionFailed):
            api.get_instructor_courses()

        with self.assertRaises(NameResolutionFailed):
            api.get_course_students(420)

    def test_retries_on_5xx(self):
        self.mock_get.side_effect = HTTPError(response=FakeResponse(503))

        api = CanvasAPI("test token", "mst.instructure.com")
        with self.assertRaises(HTTPError):
            api.get_instructor_courses()

        self.assertEqual(5, self.mock_get.call_count)


redkyn/canvas/utils.py
from redkyn.canvas import CanvasAPI
from typing import List, Dict, Tuple, Any
from redkyn.canvas.exceptions import AssignmentNotFound
import logging
import re

logger = logging.getLogger(__name__)


def lookup_canvas_ids(
    courses: List[Dict[str, Any]], canvas: CanvasAPI, hw_name: str
) -> Tuple[Dict[str, int], Dict[str, int]]:
    """
    Gets the Canvas internal course IDs for all sections of a course (if different sections have their
    own course IDs), in addition to the assignment IDs for a given assignment, within each course
    :param courses: the data containing course IDs and their sections
    :param canvas: an API object
    :param hw_name: the name of the assignment to obtain IDs for
    :return: the list of course ids keyed by the section name, and the list
    of assignment ids corresponding to each section as a dict keyed on the section
    """
    section_ids = {course["section"]: course["id"] for course in courses}
    min_name = re.search("[A-Za-z]+\d+", hw_name).group(0)
    assignment_ids = {}
    for section, course_id in section_ids.items():
        try:
            canvas_assignments = canvas.get_course_assignments(course_id, min_name)
        except Exception as e:
            print(e, type(e))
            logger.error("Failed to pull assignment list from Canvas")
            raise AssignmentNotFound
        if len(canvas_assignments) != 1:
            logger.error(
                "Could not uniquely identify Canvas assignment from name %s and section %s",
                min_name,
                section,
            )
            raise AssignmentNotFound
        assignment_ids[section] = canvas_assignments[0]["id"]
    return (section_ids, assignment_ids)


redkyn/canvas/__init__.py
import requests
from requests.exceptions import HTTPError

from urllib.parse import urljoin
import logging
import time

from redkyn.canvas.exceptions import (
    raiseAuthenticationFailed,
    raiseCourseNotFound,
    raiseStudentNotFound,
    raiseNameResolutionFailed,
)

from typing import Tuple


# Transparently use a common TLS session for each request
requests = requests.Session()


class CanvasAPI:
    REQUEST_HEADER = {"Authorization": "Bearer ", "Content-Type": "application/json"}

    def __init__(self, canvas_token: str, website_root: str):
        self.REQUEST_HEADER["Authorization"] += canvas_token

        if website_root.startswith("http://"):
            website_root = website_root[7:]

        if not website_root.startswith("https://"):
            website_root = "https://{}".format(website_root)

        self.website_root = website_root

    def _get_request(
        self, url: str, params: dict = None, attempts: int = 5
    ) -> Tuple[str, str]:
        url = urljoin(self.website_root, url)
        tries = 0
        while tries < attempts:
            try:
                r = requests.get(url, params=params, headers=self.REQUEST_HEADER)
                r.raise_for_status()
                if "Link" in r.headers:
                    return (r.json(), r.headers["Link"])
                return (r.json(), "")
            except HTTPError as e:
                if e.response is None or e.response.status_code < 500:
                    raise

                tries += 1
                if tries == attempts:
                    raise

                logging.debug(
                    "Caught %d exception in request after %d tries. Will retry %d more times.",
                    e.response.status_code,
                    tries,
                    attempts - tries,
                    exc_info=True,
                )
                time.sleep(0.5 * 2 ** (tries - 1))

    def _get_all_pages(self, url: str, params: dict = None) -> list:
        """
        Get the full results from a query by following pagination links until the last page
        :param url: The URL for the query request
        :param params: A dictionary of parameter names and values to be sent with the request
        :return: The full list of result objects returned by the query
        """
        try:
            (result, link_header) = self._get_request(url, params)
            count = len(result)
            page = 1
            while 'rel="next"' in link_header:
                page += 1

                params["per_page"] = count
                params["page"] = page
                (next_result, link_header) = self._get_request(url, params)

                result.extend(next_result)

            return result

        except HTTPError as e:
            raiseNameResolutionFailed(e)

            logging.debug("Got HTTP status %d\n", e.response.status_code)
            raiseAuthenticationFailed(e)
            raise

    def _put_request(
        self, url: str, params: dict = None, attempts: int = 5
    ) -> Tuple[str, str]:
        url = urljoin(self.website_root, url)
        tries = 0
        while tries < attempts:
            try:
                r = requests.put(url, params=params, headers=self.REQUEST_HEADER)
                r.raise_for_status()
                if "Link" in r.headers:
                    return (r.json(), r.headers["Link"])
                return (r.json(), "")
            except HTTPError as e:
                if e.response is None or e.response.status_code < 500:
                    raise

                tries += 1
                if tries == attempts:
                    raise

                logging.debug(
                    "Caught %d exception in request after %d tries. Will retry %d more times.",
                    e.response.status_code,
                    tries,
                    attempts - tries,
                    exc_info=True,
                )
                time.sleep(0.5 * 2 ** (tries - 1))

    def get_instructor_courses(self):
        get = lambda x: self._get_all_pages(
            "/api/v1/courses", {"enrollment_type": x, "state": ["available"]}
        )
        result = get("teacher")
        result.extend(get("ta"))
        result.extend(get("grader"))
        return result

    def get_course_students(self, course_id: str):
        try:
            params = {
                "per_page": 50,
                "enrollment_type": ["student"],
                "enrollment_state": ["active"],
            }
            result = self._get_all_pages("/api/v1/courses/%s/users" % course_id, params)
            return result

        except HTTPError as e:
            raiseCourseNotFound(e)
            raise

    def get_course_assignments(self, course_id: str, search_term: str = ""):
        try:
            params = {"per_page": 50}
            if search_term:
                params["search_term"] = search_term
            result = self._get_all_pages(
                "/api/v1/courses/%s/assignments" % course_id, params
            )
            return result

        except HTTPError as e:
            raiseCourseNotFound(e)
            raise

    def put_assignment_submission(
        self, course_id: str, assignment_id: str, student_id: str, score: float
    ):
        try:
            params = {"submission[posted_grade]": score}
            result = self._put_request(
                "/api/v1/courses/%s/assignments/%s/submissions/%s"
                % (course_id, assignment_id, student_id),
                params,
            )
            return result

        except HTTPError as e:
            raiseStudentNotFound(e)
            raise


redkyn/canvas/exceptions.py
from redkyn.exceptions import RedkynError
from requests.exceptions import HTTPError


def raiseAuthenticationFailed(err: HTTPError):
    if err.response.status_code != 401:
        return

    raise AuthenticationFailed(err)


class AuthenticationFailed(RedkynError):
    pass


def raiseCourseNotFound(err: HTTPError):
    if err.response.status_code != 404:
        return

    raise CourseNotFound(err)


def raiseStudentNotFound(err: HTTPError):
    if err.response.status_code != 404:
        return

    raise StudentNotFound(err)


class CourseNotFound(RedkynError):
    pass


class AssignmentNotFound(RedkynError):
    pass


class StudentNotFound(RedkynError):
    pass


def raiseNameResolutionFailed(err: HTTPError):
    if err.response is not None:
        return

    if err.request is not None:
        return

    if "Name or service not known" not in str(err):
        return

    raise NameResolutionFailed(err)


class NameResolutionFailed(RedkynError):
    pass


redkyn/canvas.py
import requests
from requests.exceptions import HTTPError

from urllib.parse import urljoin
import logging
import time

from typing import Tuple


# Transparently use a common TLS session for each request
requests = requests.Session()

class CanvasAPI:
    REQUEST_HEADER = {
        "Authorization": "Bearer ",
        "Content-Type": "application/json"
    }

    def __init__(self, canvas_token: str, website_root: str):
        self.REQUEST_HEADER['Authorization'] += canvas_token

        if website_root.startswith("http://"):
            website_root = website_root[7:]

        if not website_root.startswith("https://"):
            website_root = "https://{}".format(website_root)

        self.website_root = website_root

    def _get_request(self, url: str, params: dict = None, attempts: int = 5) -> Tuple[str, str]:
        url = urljoin(self.website_root, url)
        tries = 0
        while tries < attempts:
            try:
                r = requests.get(url, params=params, headers=self.REQUEST_HEADER)
                r.raise_for_status()
                return (r.json(), r.headers["Link"])
            except HTTPError as e:
                if e.response is None or e.response.status_code < 500:
                    raise

                tries += 1
                if tries == attempts:
                    raise

                logging.debug("Caught %d exception in request after %d tries. Will retry %d more times.",
                              e.response.status_code, tries, attempts - tries, exc_info=True)
                time.sleep(0.5 * 2 ** (tries - 1))

    def _get_all_pages(self, url: str, params: dict = None) -> list:
        """
        Get the full results from a query by following pagination links until the last page
        :param url: The URL for the query request
        :param params: A dictionary of parameter names and values to be sent with the request
        :return: The full list of result objects returned by the query
        """
        try:
            (result, link_header) = self._get_request(url, params)

            count = len(result)
            page = 1
            while 'rel="next"' in link_header:
                page += 1

                params['per_page'] = count
                params['page'] = page
                (next_result, link_header) = self._get_request(url, params)

                result.extend(next_result)

            return result

        except HTTPError as e:
            raiseNameResolutionFailed(e)

            logging.debug("Got HTTP status %d\n", e.response.status_code)
            raiseAuthenticationFailed(e)
            raise

    def get_instructor_courses(self):
        get = lambda x: self._get_all_pages('/api/v1/courses',
                                            {'enrollment_type': x, 'state': ['available']})
        result = get('teacher')
        result.extend(get('ta'))
        result.extend(get('grader'))
        return result

    def get_course_students(self, course_id: str):
        try:
            params = {
                'per_page': 50,
                'enrollment_type': ['student'],
                'enrollment_state': ['active']
            }
            result = self._get_all_pages('/api/v1/courses/%s/users' % course_id, params)
            return result

        except HTTPError as e:
            raiseCourseNotFound(e)
            raise



def raiseAuthenticationFailed(err: HTTPError):
    if err.response.status_code != 401:
        return

    raise AuthenticationFailed(err)

class AuthenticationFailed(Exception):
    pass


def raiseCourseNotFound(err: HTTPError):
    if err.response.status_code != 404:
        return

    raise CourseNotFound(err)

class CourseNotFound(Exception):
    pass


def raiseNameResolutionFailed(err: HTTPError):
    if err.response is not None:
        return

    if err.request is not None:
        return

    print(str(err))
    if "Name or service not known" not in str(err):
        return

    raise NameResolutionFailed(err)

class NameResolutionFailed(Exception):
    pass
